lib/mayaUsd/fileio/shaderReaderRegistry.h

```diff
@@ -0,0 +1,137 @@
+    typedef std::function<UsdMayaPrimReaderSharedPtr(const UsdMayaPrimReaderArgs&)> ReaderFactoryFn;
+
+    /// Reader function, i.e. a function that reads a prim. This is the
+    /// signature of the function declared in the PXRUSDMAYA_DEFINE_READER
+    /// macro.
+    typedef std::function<bool(const UsdMayaPrimReaderArgs&, UsdMayaPrimReaderContext*)> ReaderFn;
+
+    /// \brief Register \p fn as a factory function providing a
+    /// UsdMayaShaderReader subclass that can be used to read \p usdInfoId.
+    /// If you can't provide a valid UsdMayaShaderReader for the given arguments,
+    /// return a null pointer from the factory function \p fn.
+    ///
+    /// Example for registering a reader factory in your custom plugin:
+    /// \code{.cpp}
+    /// class MyReader : public UsdMayaShaderReader {
+    ///     static UsdMayaPrimReaderSharedPtr Create(
+    ///             const UsdMayaPrimReaderArgs&);
+    /// };
+    /// TF_REGISTRY_FUNCTION_WITH_TAG(UsdMayaShaderReaderRegistry, MyReader) {
+    ///     UsdMayaShaderReaderRegistry::Register("myCustomInfoId",
+    ///             MyReader::Create);
+    /// }
+    /// \endcode
+    MAYAUSD_CORE_PUBLIC
+    static void Register(const std::string& usdInfoId, ReaderFactoryFn fn);
+
+    /// \brief Wraps \p fn in a ReaderFactoryFn and registers the wrapped
+    /// function as a prim reader provider.
+    /// This is a helper method for the macro PXRUSDMAYA_DEFINE_SHADER_READER;
+    /// you probably want to use PXRUSDMAYA_DEFINE_SHADER_READER directly
+    /// instead.
+    MAYAUSD_CORE_PUBLIC
+    static void RegisterRaw(const std::string& usdInfoId, ReaderFn fn);
+
+    /// \brief Finds a reader if one exists for \p usdInfoId.
+    ///
+    /// If there is no reader plugin for \p usdInfoId, returns nullptr.
+    MAYAUSD_CORE_PUBLIC
+    static ReaderFactoryFn Find(const std::string& usdInfoId);
+};
+
+// Note, TF_REGISTRY_FUNCTION_WITH_TAG needs a type to register with so we
+// create a dummy struct in the macro.
+
+/// \brief Registers a pre-existing reader class for the given USD info:id;
+/// the reader class should be a subclass of UsdMayaShaderReader with a
+/// constructor that takes <tt>(const UsdMayaPrimReaderArgs& readerArgs)</tt>
+/// as argument.
+///
+/// Example:
+/// \code{.cpp}
+/// class MyReader : public UsdMayaPrimReader {
+///     MyReader(
+///             const MFnDependencyNode& depNodeFn,
+///             const SdfPath& usdPath,
+///             UsdMayaReadeJobContext& jobCtx) {
+///         // ...
+///     }
+/// };
+/// PXRUSDMAYA_REGISTER_WRITER(myCustomMayaNode, MyReader);
+/// \endcode
+#define PXRUSDMAYA_REGISTER_SHADER_READER(usdInfoId, readerClass)                         \
+    TF_REGISTRY_FUNCTION_WITH_TAG(UsdMayaShaderReaderRegistry, usdInfoId##_##readerClass) \
+    {                                                                                     \
+        UsdMayaShaderReaderRegistry::Register(                                            \
```

Comment:
This will add a more helpful error that will help understand the unhelpful error messages about not being able to convert the lambda to a ReaderFactoryFn (because the lambda could not return a UsdMayaPrimReaderSharedPtr),